pkg/replicaset/replicaset.go

```diff
@@ -0,0 +1,449 @@
+package replicaset
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// Builder provides struct for replicaset object containing connection to the cluster and the replicaset definitions.
+type Builder struct {
+	// Replicaset definition. Used to create a replicaset object.
+	Definition *appsv1.ReplicaSet
+	// Created replicaset object.
+	Object *appsv1.ReplicaSet
+	// Used in functions that define or mutate replicaset definition. errorMsg is processed before the replicaset
+	// object is created.
+	errorMsg  string
+	apiClient goclient.Client
+}
+
+// AdditionalOptions additional options for replicaset object.
+type AdditionalOptions func(builder *Builder) (*Builder, error)
+
+var retryInterval = time.Second * 3
+
+// NewBuilder creates a new instance of Builder.
+func NewBuilder(
+	apiClient *clients.Settings, name, nsname string, containerSpec corev1.Container) *Builder {
+	glog.V(100).Infof(
+		"Initializing new replicaset structure with the following params: "+
+			"name: %s, namespace: %s, containerSpec %v",
+		name, nsname, containerSpec)
+
+	builder := Builder{
+		apiClient: apiClient.Client,
+		Definition: &appsv1.ReplicaSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	builder.WithAdditionalContainerSpecs([]corev1.Container{containerSpec})
+
+	if name == "" {
+		glog.V(100).Infof("The name of the replicaset is empty")
+
+		builder.errorMsg = "replicaset 'name' cannot be empty"
+	}
+
+	if nsname == "" {
```

Comment:
I'd also check that `containerSpec` is not `nil`

Reply:
but containerSpec can be nil. We could decide to add containers config after using WithAdditionalContainerSpecs method.
To be more clear I remove this option from the new builder at all